DOC: Add section on how to start contributing (#25214)
matplotlib/matplotlib: doc/devel/contributing.rst

Below, you can find a number of ways to contribute, and how to connect with the
Matplotlib community.

.. _start-contributing:

> Is there a recommended way to get started with contributing?
> ------------------------------------------------------------

Reviewer: ```suggestion
Getting started with contributing
---------------------------------
```

There is no pre-defined pathway for new contributors; you can look at existing
issue and pull request discussions, and follow the discussions during code

> review. Or you can deep-dive into a subset of code-base to understand what is

Reviewer 2: ```suggestion
review. Or you can deep-dive into a subset of the code-base to understand what is
```

> review. Or you can deep-dive into a subset of code-base to understand what is

Author: Good call. I rephrased

> There is no pre-defined pathway for new contributors, but in practical terms you
> can look at existing issues and pull request discussions, as well as subsets
> of the Matplotlib source code. You can learn a lot from the observations and
> corrections that come up in code review, for example.

Reviewer: ```suggestion
There is no pre-defined pathway for new contributors; you
can look at existing issue and pull request discussions, and follow the discussions
during code review.   Or you can deep-dive into a subset of code-base to understand 
what is going on.
```

> There are a few very typical new contributor profiles:

Reviewer: ```suggestion
There are a few typical new contributor profiles:
```

* **You are a Matplotlib user, and you see a bug, a potential improvement, or
  something that annoys you, and you can fix it.**

  This is how a lot of people approach open-source development. You can search
  our issue tracker for an existing issue that describes your problem or open a
  new issue to inform us of the problem you observed and discuss the best

>   approach to fix it. If your contributions would not be captured on GitHub

Reviewer 2: I'm thinking that it is not really required to open an issue before fixing it? I often see people opening an issue the same minute they push a PR. While that can be relevant sometimes, I do not think that is the norm nor what we want to encourage? The wording above *may* lead to that impression, although I' not really sure if it is worthwhile modifying in any way...

Reviewer 2: Fair point. I have an opposite experience in SymPy, but it doesn't seem to be that much of a problem here admittedly.

  (social media, communication, educational content), you can also reach out to
  us on gitter_, `Discourse <https://discourse.matplotlib.org/>`__ or attend any
  of our `community meetings <https://scientific-python.org/calendars>`__.

* **You are not a regular Matplotlib user but a domain expert: you know about
  visualization, 3D plotting, design, technical writing, statistics, or some

>   other field where Matplotlib could be improved.**

Reviewer 2: ```suggestion
  other field where Matplotlib can be improved.**
```

Reviewer 3: I find the existing text to read better.

  In this case, even if you need help with programming skills or don't know
  where to start, you have a focus on a specific application and domain and can
  start there. In this case, maintainers can help you figure out the best

>   implementation; you can open a pull request with a starting point, and we'll
>   be happy to discuss technical approaches. If you prefer, you can use the

Reviewer: ```suggestion
  implementation; you can open an issue or pull request with a starting point, and we'll
  be happy to discuss technical approaches. If you prefer, you can use the
```

  `GitHub functionality for "draft" pull requests <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft>`__

>   and request early feedback on whatever you are working on.

Reviewer 2: My impression is that it is much easier to get feedback if it is not in draft mode... But I agree that it would be nice if it worked as above.

Author: Yeah, I think that is fair. How about this version?

---------------------

The incubator is our non-public communication channel for new contributors. It
is a private gitter_ (chat) room moderated by core Matplotlib developers where
you can get guidance and support for your first few PRs. It's a place where you
can ask questions about anything: how to use git, GitHub, how our PR review
process works, technical questions about the code, what makes for good

> documentation or a blog post, how to get involved in community work, or get a

Reviewer 2: blog post?

Author: Yes there is a blog 😄 https://blog.scientific-python.org/tags/matplotlib/

PR is an outline, unlikely to work, or stalled, and the original author is
unresponsive, feel free to open a new PR referencing the old one.

> .. _start-contributing:

Author: Let me know if this works better!

Is there a recommended way to get started with contributions?
-------------------------------------------------------------

There is no pre-defined pathway for new contributors, but in practical terms -
looking at subsets of the Matplotlib source code is reasonable, as is looking at

> existing issue and pull request discussions. You can learn a lot from the

Author: Fair point - let me know if this rewording is better

There are a few very typical new contributor profiles:

* **You are a Matplotlib user, and you see a bug, a potential improvement, or
  something that annoys you, and you can fix it.**

  This is how a lot of people approach open-source development. You can search
  our issue tracker for an existing issue that describes your problem or open a
  new issue to inform us of the problem you observed and discuss the best
  approach to fix it. If your contributions would not be captured on GitHub
  (social media, communication, educational content), you can also reach out to
  us on gitter_, `Discourse <https://discourse.matplotlib.org/>`__ or attend any
  of our `community meetings <https://scientific-python.org/calendars>`__.

* **You are not a regular Matplotlib user but a domain expert (you know about
  visualization, 3D plotting, design, technical writing, statistics, or some
  other field where Matplotlib could be improved).**

  In this case, even if you need help with programming skills or don't know
  where to start, you have a focus on a specific application and domain and can
  start there. In this case, maintainers can help you figure out the best

>   implementation; you can open a (draft or not) pull request with a starting

Author: Makes sense - linked to the docs, let me know if this is ok

new_contributor_meeting>`_.
We recommend joining the meeting to clarify any doubts, or lingering
questions you might have, and to get to know a few of the people behind the

> GitHub handles 😉. You can reach out to us on gitter_ for any clarifications or
> suggestions. We <3 feedback!

Reviewer 2: ```suggestion
GitHub handles 😉. You can reach out to us on gitter_ for any clarifications or
suggestions. We ❤ feedback!
```
